Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,156 @@
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Math.NumberTheory.Primes.Factorisation.LinearAlgebra
+  ( SBVector(..)
+  , DBVector(..)
+  , SBMatrix(..)
+  , dot
+  , mult
+  , size
+  , linearSolve
+  ) where
+
+import qualified Data.List as L
+import qualified Data.Vector.Unboxed as U
+import qualified Data.Vector.Sized as SV
+import qualified Data.Vector.Unboxed.Sized as SU
+import qualified Data.Vector.Unboxed.Mutable as MU
+import qualified Data.Vector.Unboxed.Mutable.Sized as SMU
+import Control.Monad.ST
+import Data.Semigroup()
+import System.Random
+import System.IO.Unsafe
+import System.CPUTime
+import Data.Bit
+import Data.Bits
+import Data.Foldable
+import GHC.TypeNats hiding (Mod)
+import Data.Mod.Word
+import Unsafe.Coerce
+import Data.Maybe
+
+-- Sparse Binary Vector
+newtype SBVector (k :: Nat) = SBVector { unSBVector :: U.Vector (Mod k) }
+
+-- Dense Binary Vector
+newtype DBVector (k :: Nat) = DBVector { unDBVector :: SU.Vector k Bit }
+  deriving (Eq, Show)
+
+-- Sparse Binary Matrix
+newtype SBMatrix (k :: Nat) = SBMatrix { unSBMatrix :: SV.Vector k (SBVector k) }
+
+instance KnownNat k => Semigroup (DBVector k) where
+  DBVector v1 <> DBVector v2 = DBVector $ SU.withVectorUnsafe (zipBits xor (SU.fromSized v1)) v2
+
+instance KnownNat k => Monoid (DBVector k) where
+  mempty = DBVector $ SU.replicate (Bit False)
+
+listBits' :: KnownNat k => DBVector k -> [Mod k]
+listBits' = unsafeCoerce listBits
+
+flipBit' :: KnownNat k => SMU.MVector k s Bit -> Mod k -> ST s ()
+flipBit' = unsafeCoerce (unsafeFlipBit :: MU.MVector s Bit -> Int -> ST s ())
+
+index' :: KnownNat k => SBMatrix k -> Mod k -> SBVector k
+index' = unsafeCoerce SV.index
+
+dot :: KnownNat k => DBVector k -> DBVector k -> Bit
+dot (DBVector v1) (DBVector v2) = Bit $ odd . countBits $ zipBits (.&.) (SU.fromSized v1) (SU.fromSized v2)
+
+mult :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k
+mult matrix vector = runST $ do
+  vs <- SMU.new
+  traverse_ (traverse_ (flipBit' vs) . U.toList . unSBVector . (matrix `index'`)) $ listBits' vector
+  ws <- SU.unsafeFreeze vs
+  pure $ DBVector ws
+
+size :: KnownNat k => SBMatrix k -> Int
+size (SBMatrix m) = SV.length m
```

Comment:
This function feels redundant to me, you can use `natVal` to demote size of the matrix from the type level to the term level.